homeassistant/components/automation/__init__.py

```diff
@@ -248,41 +233,60 @@ def is_on(self) -> bool:
         """Return True if entity is on."""
         return self._enabled
 
-    def turn_on(self, **kwargs) -> None:
-        """Turn the entity on."""
-        if self._enabled:
-            return
-
-        self._detach_triggers = self._attach_triggers(self.trigger)
-        self._enabled = True
-        self.update_ha_state()
+    @asyncio.coroutine
+    def async_turn_on(self, **kwargs) -> None:
+        """Turn the entity on and update the state."""
+        yield from self.async_enable()
+        yield from self.async_update_ha_state()
 
-    def turn_off(self, **kwargs) -> None:
+    @asyncio.coroutine
+    def async_turn_off(self, **kwargs) -> None:
         """Turn the entity off."""
         if not self._enabled:
             return
 
-        self._detach_triggers()
-        self._detach_triggers = None
+        self._async_detach_triggers()
+        self._async_detach_triggers = None
         self._enabled = False
-        self.update_ha_state()
+        yield from self.async_update_ha_state()
+
+    @asyncio.coroutine
+    def async_toggle(self):
+        """Toggle the state of the entity."""
+        if self._enabled:
+            yield from self.async_turn_off()
+        else:
+            yield from self.async_turn_on()
```

Comment:
Does the caller need this to occur before returning? If this is fired off as a bus event, probably not.